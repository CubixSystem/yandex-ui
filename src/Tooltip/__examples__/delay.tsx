import React from 'react';
import { Button } from '@yandex-lego/components/Button/desktop/bundle';
import { TooltipStateful } from '@yandex-lego/components/Tooltip/desktop/bundle';

export const Delay = () => (
    <div style={{ display: 'grid', gridTemplateColumns: 'repeat(2, max-content)', gap: '32px' }}>
        <TooltipStateful view="default" size="m" hasTail content="Opened after 500ms" trigger="hover" openDelay={500}>
            <Button view="default" size="m">
                Delay Open - 500ms
            </Button>
        </TooltipStateful>
        <TooltipStateful view="default" size="m" hasTail content="Closed after 500ms" trigger="hover" closeDelay={500}>
            <Button view="default" size="m">
                Delay Close - 500ms
            </Button>
        </TooltipStateful>
    </div>
);
