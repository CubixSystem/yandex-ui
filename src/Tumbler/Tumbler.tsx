import React, {
    FC,
    ChangeEventHandler,
    useCallback,
    useRef,
    RefObject,
    MouseEventHandler,
    useMemo,
    KeyboardEventHandler,
    KeyboardEvent,
    ReactNode,
    FocusEventHandler,
    useEffect,
} from 'react';
import { cn } from '@bem-react/classname';

import { Keys, isKeyCode } from '../lib/keyboard';
import { mergeRefs } from '../lib/mergeRefs';
import { useUniqId } from '../useUniqId';
import { TumblerLabel as Label } from './Label/Tumbler-Label';
import { TumblerButton as Button } from './Button/Tumbler-Button';
import { TumblerControl as Control } from './Control/Tumbler-Control';
import './Tumbler.css';

export const CHECKED_KEYS = [Keys.RIGHT, Keys.UP];
export const UNCHECKED_KEYS = [Keys.LEFT, Keys.DOWN];

export type TumblerProps = {
    /**
     * Состояние переключателя
     */
    checked: boolean;

    /**
     * Дополнительный класс у корневого DOM-элемента
     */
    className?: string;

    /**
     * Ссылка на нативный DOM-элемент нативного инпута
     */
    controlRef?: RefObject<HTMLInputElement>;

    /**
     * Неактивное состояние переключателя
     */
    disabled?: boolean;

    /**
     * Уникальный id компонента
     */
    id?: string;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLSpanElement>;

    /**
     * Подпись после переключателя
     */
    labelAfter?: ReactNode;

    /**
     * Подпись перед переключателем
     */
    labelBefore?: ReactNode;

    /**
     * Имя переключателя в форме
     */
    name?: string;

    /**
     * Обработчик, который вызывается при потере фокуса переключателем
     */
    onBlur?: FocusEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при изменении состояния
     */
    onChange: ChangeEventHandler<HTMLInputElement>;

    /**
     * Обработчик, который вызывается при нажатии на переключатель
     */
    onClick?: MouseEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при получении фокуса переключателем
     */
    onFocus?: FocusEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при нажатии клавиши на клавиатуре (при этом переключатель должен быть в фокусе)
     */
    onKeyDown?: KeyboardEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при отжатии клавиши на клавиатуре (при этом переключатель должен быть в фокусе)
     */
    onKeyUp?: KeyboardEventHandler<HTMLButtonElement>;

    /**
     * Целое число, определяющее должен ли переключатель участвовать
     * в последовательной навигации по всей странице с помощью клавиатуры
     */
    tabIndex?: number;

    /**
     * Всплывающая подсказка
     */
    title?: string;

    /**
     * Устанавливает фокус в компонент при монтировании
     */
    autoFocus?: boolean;

    /**
     * Устанавливает в компоненте обязательное состояние
     */
    required?: boolean;
};

export const cnTumbler = cn('Tumbler');

/**
 * Компонент, предназначенный для создания переключателя.
 *
 * @param {TumblerProps} props Свойства компонента.
 */
export const Tumbler: FC<TumblerProps> = ({
    checked,
    className,
    controlRef: htmlControlRef,
    disabled,
    // eslint-disable-next-line react-hooks/rules-of-hooks
    id = useUniqId(),
    innerRef,
    labelAfter,
    labelBefore,
    name,
    onBlur,
    onChange,
    onClick: htmlOnClick,
    onFocus,
    onKeyDown: htmlOnKeyDown,
    onKeyUp,
    tabIndex,
    title,
    autoFocus,
    required,
}) => {
    const buttonRef = useRef<HTMLButtonElement>(null);
    const controlRef = useRef<HTMLInputElement>(null);

    const controlId = `control-${id}`;
    const beforeLabelId = `before-label-${id}`;
    const afterLabelId = `after-label-${id}`;

    useEffect(() => {
        if (autoFocus && buttonRef.current !== null) {
            buttonRef.current.focus();
        }
        // TODO ISL-10952: разобраться почему не все необходимые переменные указаны в deps и убрать игнор
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const labelledById = useMemo(() => {
        if (labelBefore && labelAfter) return checked ? beforeLabelId : afterLabelId;
        if (labelBefore) return beforeLabelId;
        if (labelAfter) return afterLabelId;
        return undefined;
    }, [checked, labelBefore, labelAfter, beforeLabelId, afterLabelId]);

    // prettier-ignore
    const onButtonClick = useCallback((event) => {
        if (controlRef.current !== null) {
            controlRef.current.click();
        }
        if (buttonRef.current !== null) {
            buttonRef.current.focus();
        }
        if (htmlOnClick !== undefined) {
            htmlOnClick(event);
        }
    }, [htmlOnClick]);

    // prettier-ignore
    const onLabelClick = useCallback(() => {
        if (buttonRef.current !== null) {
            buttonRef.current.focus();
        }
    }, []);

    // prettier-ignore
    const onKeyDown = useCallback((event: KeyboardEvent<HTMLButtonElement>) => {
        if (isKeyCode(event.keyCode, [Keys.ENTER, ...UNCHECKED_KEYS, ...CHECKED_KEYS])) {
            event.preventDefault();
        }
        const shouldForceChange =
            (checked && isKeyCode(event.keyCode, UNCHECKED_KEYS)) ||
            (!checked && isKeyCode(event.keyCode, CHECKED_KEYS));
        if (shouldForceChange && controlRef.current !== null) {
            controlRef.current.click();
        }
        if (htmlOnKeyDown !== undefined) {
            htmlOnKeyDown(event);
        }
    }, [checked, htmlOnKeyDown]);

    return (
        <span ref={innerRef} className={cnTumbler(null, [className])} id={id} title={title} aria-disabled={disabled}>
            {labelBefore && (
                <Label
                    disabled={Boolean(!checked && labelBefore && labelAfter)}
                    htmlFor={controlId}
                    id={beforeLabelId}
                    onClick={onLabelClick}
                >
                    {labelBefore}
                </Label>
            )}
            <Button
                tabIndex={disabled ? -1 : tabIndex}
                innerRef={buttonRef}
                checked={checked}
                onKeyDown={onKeyDown}
                onKeyUp={onKeyUp}
                onBlur={onBlur}
                onFocus={onFocus}
                onClick={onButtonClick}
                labelledBy={labelledById}
                required={required}
            />
            <Control
                id={controlId}
                name={name}
                checked={checked}
                innerRef={mergeRefs(controlRef, htmlControlRef)}
                onChange={onChange}
                required={required}
            />
            {labelAfter && (
                <Label
                    disabled={Boolean(checked && labelBefore && labelAfter)}
                    htmlFor={controlId}
                    id={afterLabelId}
                    onClick={onLabelClick}
                >
                    {labelAfter}
                </Label>
            )}
        </span>
    );
};

Tumbler.displayName = 'Tumbler';
