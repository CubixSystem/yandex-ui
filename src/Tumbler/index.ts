export * from './Tumbler';
export * from './_size/Tumbler_size_l';
export * from './_size/Tumbler_size_m';
export * from './_size/Tumbler_size_s';
export * from './_view/Tumbler_view_default';
