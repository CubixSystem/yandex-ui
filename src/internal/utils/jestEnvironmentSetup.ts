/**
 * Возвращает DOM-узел для монтирования компонента.
 */
export const getRoot = (): HTMLElement | null => {
    return document.getElementById('root');
};

/**
 * Хелпер, создающий корневой контейнер для монтирования,
 * необходим для того, чтобы избежать ошибок при монтировании в enzyme.
 */
export const jestEnvironmentSetup = (): void => {
    // eslint-disable-next-line no-undef
    beforeEach(() => {
        const node = document.createElement('div');
        node.setAttribute('id', 'root');
        document.body.appendChild(node);
    });

    // eslint-disable-next-line no-undef
    afterEach(() => {
        const node = getRoot();
        if (node !== null) {
            document.body.removeChild(node);
        }
    });
};
