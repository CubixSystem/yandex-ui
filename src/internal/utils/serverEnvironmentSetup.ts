/**
 * Хелпер, который симулирует серверное окружение во время запуска тестов.
 */
export const serverEnvironmentSetup = () => {
    const prevWindow = (global as any).window;
    const prevDocument = (global as any).document;
    const prevRAF = (global as any).requestAnimationFrame;

    // eslint-disable-next-line no-undef
    beforeAll(() => {
        delete (global as any).window;
        delete (global as any).document;
        delete (global as any).requestAnimationFrame;
    });

    // eslint-disable-next-line no-undef
    afterAll(() => {
        (global as any).window = prevWindow;
        (global as any).document = prevDocument;
        (global as any).requestAnimationFrame = prevRAF;
    });
};
