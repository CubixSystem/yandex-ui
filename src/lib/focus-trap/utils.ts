import { getTabbables } from './tabbable';

const FOCUS_TRAP_GUARD_ATTR = 'focusTrapGuard';

interface ScrollableElement {
    element: HTMLElement;
    scrollTop: number;
    scrollLeft: number;
}

let supportsPreventScrollCached: boolean | null = null;
function supportsPreventScroll() {
    if (supportsPreventScrollCached === null) {
        supportsPreventScrollCached = false;

        try {
            const element = document.createElement('div');

            element.focus({
                get preventScroll() {
                    supportsPreventScrollCached = true;
                    return true;
                },
            });
        } catch (e) {
            // ignore
        }
    }

    return supportsPreventScrollCached;
}

function getScrollableElements(element: HTMLElement): ScrollableElement[] {
    const scrollableElements: ScrollableElement[] = [];
    const rootScrollingElement = document.scrollingElement || document.documentElement;

    let parent = element.parentNode;
    while (parent instanceof HTMLElement && parent !== rootScrollingElement) {
        if (parent.offsetHeight < parent.scrollHeight || parent.offsetWidth < parent.scrollWidth) {
            scrollableElements.push({
                element: parent,
                scrollTop: parent.scrollTop,
                scrollLeft: parent.scrollLeft,
            });
        }
        parent = parent.parentNode;
    }

    if (rootScrollingElement instanceof HTMLElement) {
        scrollableElements.push({
            element: rootScrollingElement,
            scrollTop: rootScrollingElement.scrollTop,
            scrollLeft: rootScrollingElement.scrollLeft,
        });
    }

    return scrollableElements;
}

function restoreScrollPosition(scrollableElements: ScrollableElement[]) {
    for (const { element, scrollTop, scrollLeft } of scrollableElements) {
        element.scrollTop = scrollTop;
        element.scrollLeft = scrollLeft;
    }
}

export function getActiveElement() {
    return document.activeElement as HTMLElement | null;
}

export function isElementInScope(scope: Element, element: Element) {
    return scope.contains(element);
}

export function isFocusFree() {
    return getActiveElement() === document.body;
}

export function createGuard(tabIndex: number) {
    const element = document.createElement('span');
    element.tabIndex = tabIndex;
    Object.assign(element.style, {
        width: '1px',
        height: 0,
        padding: 0,
        overflow: 'hidden',
        position: 'fixed',
        top: '1px',
        left: '1px',
    });
    element.setAttribute('aria-hidden', 'true');
    element.dataset[FOCUS_TRAP_GUARD_ATTR] = 'true';

    return element;
}

export function isGuard(node: HTMLElement) {
    return Boolean(node.dataset[FOCUS_TRAP_GUARD_ATTR]);
}

export function removeElementFromDocument(element: Element) {
    if (!element.parentNode) {
        return;
    }

    element.parentNode.removeChild(element);
}

function focusWithoutScrolling(element: HTMLElement) {
    if (supportsPreventScroll()) {
        element.focus({ preventScroll: true });
    } else {
        const scrollableElements = getScrollableElements(element);
        element.focus();
        restoreScrollPosition(scrollableElements);
    }
}

export function focusElement(element?: HTMLElement | null, preventScroll?: boolean) {
    if (!element) {
        return;
    }

    try {
        if (preventScroll) {
            focusWithoutScrolling(element);
        } else {
            element.focus();
        }
    } catch (e) {
        // ignore
    }
}

export function focusFirstIn(scope: HTMLElement) {
    const [first] = getTabbables(scope, (node) => !isGuard(node));

    focusElement(first);
}

export function getNextTabbableIndex(current: number, count: number, step: number) {
    if (current === -1) {
        return 0;
    }

    return count > 0 ? (count + current + step) % count : -1;
}
