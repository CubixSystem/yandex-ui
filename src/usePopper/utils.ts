import { Placement, BasePlacement, Offsets, top, left, right } from '@popperjs/core';

export function getLayoutRect(element: HTMLElement) {
    return {
        x: element.offsetLeft,
        y: element.offsetTop,
        width: element.offsetWidth,
        height: element.offsetHeight,
    };
}

export default function getBasePlacement(placement: Placement): BasePlacement {
    return placement.split('-')[0] as any;
}

export function distanceAndSkiddingToXY(placement: Placement, offset: [number, number]): Offsets {
    const basePlacement = getBasePlacement(placement);
    const invertDistance = [left, top].indexOf(basePlacement as any) >= 0 ? -1 : 1;

    const skidding = offset[0] || 0;
    const distance = (offset[1] || 0) * invertDistance;

    return [left, right].indexOf(basePlacement as any) >= 0
        ? { x: distance, y: skidding }
        : { x: skidding, y: distance };
}
