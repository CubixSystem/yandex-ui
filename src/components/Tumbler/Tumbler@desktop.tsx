import React, { FC, useCallback, KeyboardEvent } from 'react';

import '../../polyfills/pointerfocus';
import { forceUtilityFocus } from '../../lib/pointerfocus';
import { Keys, isKeyCode } from '../../lib/keyboard';
import { TumblerProps, Tumbler as TumblerCommon, CHECKED_KEYS, UNCHECKED_KEYS } from './Tumbler';

export * from './Tumbler';

export const Tumbler: FC<TumblerProps> = ({ checked, onKeyDown: htmlOnKeyDown, ...props }) => {
    // prettier-ignore
    const onKeyDown = useCallback((event: KeyboardEvent<HTMLButtonElement>) => {
        if (
            (!checked && isKeyCode(event.keyCode, CHECKED_KEYS)) ||
            (checked && isKeyCode(event.keyCode, UNCHECKED_KEYS)) ||
            (event.keyCode === Keys.SPACE)
        ) {
            forceUtilityFocus();
        }

        if (htmlOnKeyDown !== undefined) {
            htmlOnKeyDown(event);
        }
    }, [checked, htmlOnKeyDown]);

    return <TumblerCommon checked={checked} onKeyDown={onKeyDown} {...props} />;
};

Tumbler.displayName = TumblerCommon.displayName;
