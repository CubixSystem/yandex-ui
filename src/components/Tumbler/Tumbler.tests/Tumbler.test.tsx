import React, { ComponentType, createRef } from 'react';
import { render, mount } from 'enzyme';

import { delay } from '../../../internal/utils/delay';
import { serverEnvironmentSetup } from '../../../internal/utils/serverEnvironmentSetup';
import { jestEnvironmentSetup, getRoot } from '../../../internal/utils/jestEnvironmentSetup';

import { TumblerProps } from '../Tumbler';
import { Tumbler as TumblerDesktop } from '../Tumbler@desktop';
import { Tumbler as TumblerTouchPad } from '../Tumbler@touch-pad';
import { Tumbler as TumblerTouchPhone } from '../Tumbler@touch-phone';

// prettier-ignore
const platforms = [
    ['desktop', TumblerDesktop],
    ['touch-pad', TumblerTouchPad],
    ['touch-phone', TumblerTouchPhone],
];

const noop = () => null;

describe.each<any>(platforms)('Tumbler@%s', (_platform, Tumbler: ComponentType<TumblerProps>) => {
    jestEnvironmentSetup();

    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(
                render(
                    <Tumbler
                        checked={false}
                        onChange={noop}
                        tabIndex={1}
                        name="switch"
                        id="static-id"
                        className="unit-mixin"
                        labelBefore="labelBefore"
                        labelAfter="labelAfter"
                    />,
                ),
            ).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(
                render(
                    <Tumbler
                        checked={false}
                        onChange={noop}
                        tabIndex={1}
                        name="switch"
                        id="static-id"
                        className="unit-mixin"
                        labelBefore="labelBefore"
                        labelAfter="labelAfter"
                    />,
                ),
            ).toMatchSnapshot();
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLSpanElement>();
            mount(<Tumbler checked={false} onChange={noop} innerRef={innerRef} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('должен устанавливать ссылку на нативный DOM элемент', async () => {
            const controlRef = createRef<HTMLInputElement>();
            mount(<Tumbler checked={false} onChange={noop} controlRef={controlRef} />);
            await delay(100);
            expect(controlRef.current).not.toBe(null);
        });

        test('должен быть установлен корректный displayName', () => {
            expect(Tumbler.displayName).toBe('Tumbler');
        });

        test('должен устанавливать aria-pressed при изменении свойства checked', () => {
            const wrapper = mount(<Tumbler checked={false} onChange={noop} />);
            expect(wrapper.find('.Tumbler-Button').prop('aria-pressed')).toBe(false);
            wrapper.setProps({ checked: true });
            expect(wrapper.find('.Tumbler-Button').prop('aria-pressed')).toBe(true);
            wrapper.setProps({ checked: false });
            expect(wrapper.find('.Tumbler-Button').prop('aria-pressed')).toBe(false);
        });

        test('должен устанавливать aria-disabled при изменении свойства disabled', () => {
            const wrapper = mount(<Tumbler disabled={false} checked={false} onChange={noop} />);
            expect(wrapper.find('.Tumbler').prop('aria-disabled')).toBe(false);
            wrapper.setProps({ disabled: true });
            expect(wrapper.find('.Tumbler').prop('aria-disabled')).toBe(true);
            wrapper.setProps({ disabled: false });
            expect(wrapper.find('.Tumbler').prop('aria-disabled')).toBe(false);
        });

        test('должен устанавливать tabIndex=-1 при изменении свойства disabled', () => {
            const wrapper = mount(<Tumbler disabled={false} checked={false} onChange={noop} tabIndex={0} />);
            expect(wrapper.find('.Tumbler-Button').prop('tabIndex')).toBe(0);
            wrapper.setProps({ disabled: true });
            expect(wrapper.find('.Tumbler-Button').prop('tabIndex')).toBe(-1);
            wrapper.setProps({ disabled: false });
            expect(wrapper.find('.Tumbler-Button').prop('tabIndex')).toBe(0);
        });

        test('должен вызывать обработчик onClick при нажатии на мышку', () => {
            const onClick = jest.fn();
            const wrapper = mount(<Tumbler onClick={onClick} checked={false} onChange={noop} />);
            wrapper.find('.Tumbler-Button').simulate('click');
            expect(onClick.mock.calls.length).toBe(1);
        });

        test('должен вызывать обработчик onKeyDown при нажатии на клавиатуру', () => {
            const onKeyDown = jest.fn();
            const wrapper = mount(<Tumbler onKeyDown={onKeyDown} checked={false} onChange={noop} />);
            wrapper.find('.Tumbler-Button').simulate('keydown');
            expect(onKeyDown.mock.calls.length).toBe(1);
        });

        test('должен устанавливать фокус в кнопку после нажатия на лейбл либо кнопку', () => {
            const wrapper = mount(<Tumbler checked={false} onChange={noop} labelBefore="l" labelAfter="r" />, {
                attachTo: getRoot(),
            });
            const buttonRef = wrapper.find('.Tumbler-Button');
            const labelBeforeRef = wrapper.find('.Tumbler-Label').at(0);
            const labelAfterRef = wrapper.find('.Tumbler-Label').at(0);
            buttonRef.simulate('click');
            expect(document.activeElement).toBe(buttonRef.instance());
            labelBeforeRef.simulate('click');
            expect(document.activeElement).toBe(buttonRef.instance());
            labelAfterRef.simulate('click');
            expect(document.activeElement).toBe(buttonRef.instance());
        });
    });
});
