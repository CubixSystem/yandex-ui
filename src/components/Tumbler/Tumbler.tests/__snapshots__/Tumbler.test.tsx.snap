// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tumbler@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;
